From feeddc7052c99587633f14681569db587d04ef68 Mon Sep 17 00:00:00 2001
From: Rishi Gupta <rishgupt@codeaurora.org>
Date: Sun, 4 Apr 2021 17:20:27 +0530
Subject: [PATCH] evm: exclude data partition from verification

1. evm_update_evmxattr: we don't want to use evmkey therefore
don't calculate HMAC and update. If a new file is created
it will not have security.evm xattr. If metadata of an
existing file is modified, we don't update its security.evm
xattr value because files on root file system are supposed
to have digital signature only.

2. evm_verify_current_integrity: in evm-enforcing mode, we
want to protect files only on root file system. Therefore,
exclude all other partitions from EVM appraisal.
Further, SD card is used to save extra debugging data in
few use-cases, exclude it also.

By-default, EVM protects all files including pseudo type
also. Exclude them as well because chmod/chown operations
are done on them for various use-cases.

3. evm_inode_init_security: it is fine that a new file is
created and it doesn't have security.evm. Therefore, return
success.

4. init_evm: remove sysfs entry that informs kernel that
HMAC key is loaded. We are not using this key.

5. evm_load_x509: if the evm public key file is deleted or
tampered, it will fail to load during boot. This results
in evm protection getting disabled therefore raise fatal
error when this situation is detected.

Signed-off-by: Rishi Gupta <rishgupt@codeaurora.org>
---
 security/integrity/evm/evm_crypto.c |  3 +++
 security/integrity/evm/evm_main.c   | 38 +++++++++++++++++++++++++++++++------
 2 files changed, 35 insertions(+), 6 deletions(-)

diff --git a/security/integrity/evm/evm_crypto.c b/security/integrity/evm/evm_crypto.c
index d5c713a..d121dda 100644
--- a/security/integrity/evm/evm_crypto.c
+++ b/security/integrity/evm/evm_crypto.c
@@ -321,6 +321,9 @@ int evm_update_evmxattr(struct dentry *dentry, const char *xattr_name,
 	if (rc)
 		return -EPERM;
 
+	/* Unconditionally return success, we don't use HMAC at all */
+	return 0;
+
 	data.hdr.algo = HASH_ALGO_SHA1;
 	rc = evm_calc_hmac(dentry, xattr_name, xattr_value,
 			   xattr_value_len, &data);
diff --git a/security/integrity/evm/evm_main.c b/security/integrity/evm/evm_main.c
index e794abf..e2ff340 100644
--- a/security/integrity/evm/evm_main.c
+++ b/security/integrity/evm/evm_main.c
@@ -273,9 +273,35 @@ EXPORT_SYMBOL_GPL(evm_verifyxattr);
 static enum integrity_status evm_verify_current_integrity(struct dentry *dentry)
 {
 	struct inode *inode = d_backing_inode(dentry);
+	uuid_t rfsuuid = {.b = 0x6F, 0x2B, 0x9A, 0x5B, 0x32, 0x07, 0x4D, 0x8B, 0x84, 0x7C, 0x0F, 0x4F, 0xF9, 0xD9, 0x04, 0x85};
 
 	if (!evm_initialized || !S_ISREG(inode->i_mode) || evm_fixmode)
 		return 0;
+
+	/* 1. Exclude all UBIFS formatted partition except root file
+	 *    system partition. 0x24051905 is UBIFS magic number.
+	 * 2. Exclude external SD-card (FAT FS). */
+	switch (dentry->d_sb->s_magic) {
+		case 0x24051905:
+			if (!uuid_equal(&rfsuuid, &inode->i_sb->s_uuid))
+				return 0;
+			break;
+		case MSDOS_SUPER_MAGIC:
+			return 0;
+	}
+
+	/* Exclude pseudo filesystems. 0x62656570 is CONFIGFS magic number */
+	switch (dentry->d_sb->s_magic) {
+		case 0x62656570:
+		case TMPFS_MAGIC:
+		case SYSFS_MAGIC:
+		case RAMFS_MAGIC:
+		case CGROUP_SUPER_MAGIC:
+		case CGROUP2_SUPER_MAGIC:
+		case DEVPTS_SUPER_MAGIC:
+			return 0;
+	}
+
 	return evm_verify_hmac(dentry, NULL, NULL, 0, NULL);
 }
 
@@ -489,6 +515,10 @@ int evm_inode_init_security(struct inode *inode,
 	if (!evm_initialized || !evm_protected_xattr(lsm_xattr->name))
 		return 0;
 
+	/* If HMAC key is not available, simply return */
+	if (!(evm_initialized & EVM_INIT_HMAC))
+		return 0;
+
 	xattr_data = kzalloc(sizeof(*xattr_data), GFP_NOFS);
 	if (!xattr_data)
 		return -ENOMEM;
@@ -516,6 +546,8 @@ void __init evm_load_x509(void)
 	rc = integrity_load_x509(INTEGRITY_KEYRING_EVM, CONFIG_EVM_X509_PATH);
 	if (!rc)
 		evm_initialized |= EVM_INIT_X509;
+	else
+		panic("can't load evm public key!");
 }
 #endif
 
@@ -529,12 +561,6 @@ static int __init init_evm(void)
 	if (error)
 		return error;
 
-	error = evm_init_secfs();
-	if (error < 0) {
-		pr_info("Error registering secfs\n");
-		return error;
-	}
-
 	return 0;
 }
 
-- 
2.7.4

